ast/src/main/java/com/graphicsfuzz/common/ast/visitors/AstBuilder.java

```diff
@@ -195,9 +198,124 @@ void addTopLevelDeclaration(Declaration decl) {
+        }
+        if (expr instanceof ParenExpr) {
+          return reduce(((ParenExpr)expr).getExpr());
+        }
+        if (expr instanceof BinaryExpr) {
+          BinaryExpr bexpr = (BinaryExpr) expr;
+
+          Expr lexpr = reduce(bexpr.getLhs());
+          Expr rexpr = reduce(bexpr.getRhs());
+
+          if (!(lexpr instanceof IntConstantExpr && rexpr instanceof IntConstantExpr)) {
+            throw new RuntimeException("Unable to fold constant " + bexpr.getText());
+          }
+
+          int lval = ((IntConstantExpr) lexpr).getNumericValue();
+          int rval = ((IntConstantExpr) rexpr).getNumericValue();
+          int fval = 0;
+          switch (bexpr.getOp()) {
+            case MOD:
+              fval = lval % rval;
+              break;
+            case MUL:
+              fval = lval * rval;
+              break;
+            case DIV:
+              if (rval == 0) {
+                throw new RuntimeException("Division by zero folding constant " + bexpr.getText());
+              }
+              fval = lval / rval;
+              break;
+            case ADD:
+              fval = lval + rval;
+              break;
+            case SUB:
+              fval = lval - rval;
+              break;
+            case BAND:
+              fval = lval & rval;
+              break;
+            case BOR:
+              fval = lval | rval;
+              break;
+            case BXOR:
+              fval = lval ^ rval;
+              break;
+            case LAND:
+              fval = ((lval != 0) && (rval != 0)) ? 1 : 0;
+              break;
+            case LOR:
+              fval = ((lval != 0) || (rval != 0)) ? 1 : 0;
+              break;
+            case LXOR:
+              fval = ((lval == 0) != (rval == 0)) ? 1 : 0;
+              break;
+            case SHL:
+              fval = lval << rval;
+              break;
+            case SHR:
+              fval = lval >> rval;
+              break;
+            case LT:
+              fval = (lval < rval) ? 1 : 0;
+              break;
+            case GT:
+              fval = (lval > rval) ? 1 : 0;
+              break;
+            case LE:
+              fval = (lval <= rval) ? 1 : 0;
+              break;
+            case GE:
+              fval = (lval >= rval) ? 1 : 0;
+              break;
+            case EQ:
+              fval = (lval == rval) ? 1 : 0;
+              break;
+            case NE:
+              fval = (lval != rval) ? 1 : 0;
+              break;
+            default:
+              throw new RuntimeException("Unable to fold constant " + bexpr.getText());
+          }
+          return new IntConstantExpr(Integer.toString(fval));
+        }
+        throw new RuntimeException("Unable to fold constant " + expr.getText());
+      }
+
+      @Override
+      public void visitVariableDeclInfo(VariableDeclInfo variableDeclInfo) {
+        if (variableDeclInfo.hasArrayInfo() && variableDeclInfo.getArrayInfo().hasSize()) {
```

Comment:
I prefer to use final variables where possible in the code, and to use a functional style if possible, so I would prefer just:

variableDeclInfo.getArrayInfo().setSizeExpr(reduce(variableDeclInfo.getArrayInfo().getSizeExpr()));

or:

final Expr constantFoldedSize = reduce(variableDeclInfo.getArrayInfo().getSizeExpr());
variableDeclInfo.getArrayInfo.setSizeExpr(constantFoldedSize);

I also think if might be better to do:

if (varibleDeclInfo.hasArrayInfo()) {
  final ArrayInfo arrayInfo = variableDeclInfo.getArrayInfo();
  ...
}

to avoid looking up variableDeclInfo.getArrayInfo() so many times.